mlir/include/mlir/IR/BuiltinTypes.h

```diff
@@ -43,6 +43,10 @@ template <typename ConcreteType>
 class ValueSemantics
     : public TypeTrait::TraitBase<ConcreteType, ValueSemantics> {};
 
+/// Type trait indicating that the type is a pointer-like type.
+template <typename ConcreteType>
+class PointerLike : public TypeTrait::TraitBase<ConcreteType, PointerLike> {};
```

Comment:
I wonder if this should be an interface... we may have common methods to retrieve the address space, at least?

Reply:
As long as there are no interface methods that we would like to add (now or in the future), I would leave it as a trait.

Reply:
I'm not sure there's demand upstream for an interface, but querying the memory space and perhaps whether the pointer is opaque could be useful for downstream. Maybe CIR and flang folks might have an use.

Reply:
Given that changing this to an interface is not really intrusive, I vote in favor of adding this once a use case pops up.

Reply:
> I'm not sure there's demand upstream for an interface, but querying the memory space and perhaps whether the pointer is opaque could be useful for downstream. Maybe CIR and flang folks might have a use.

Yes, that's what I was thinking but I'm also ok with adding this on a case by case basis.